com.ibm.wala.cast/source/java/com/ibm/wala/cast/loader/CAstAbstractLoader.java

```diff
@@ -62,31 +59,24 @@ public CAstAbstractLoader(IClassHierarchy cha) {
     this(cha, null);
   }
 
+  private Set<Warning> messagesFor(ModuleEntry module) {
+    return errors.computeIfAbsent(module, (key) -> new HashSet<>());
+  }
+
   public void addMessage(ModuleEntry module, Set<Warning> message) {
-    if (! errors.containsKey(module)) {
-      errors.put(module, new HashSet<Warning>());
-    }
-    
-    errors.get(module).addAll(message);
+    messagesFor(module).addAll(message);
   }
 
   public void addMessage(ModuleEntry module, Warning message) {
-    if (! errors.containsKey(module)) {
-      errors.put(module, new HashSet<Warning>());
-    }
-    
-    errors.get(module).add(message);
+    messagesFor(module).add(message);
   }
 
   private Iterator<ModuleEntry> getMessages(final byte severity) {
-    return new MapIterator<>(new FilterIterator<>(errors.entrySet().iterator(), o -> {
-       for(Warning w : o.getValue()) {
-         if (w.getLevel() == severity) {
-           return true;
-         }
-       }
-       return false;
-    }), Map.Entry::getKey);
+    return errors.entrySet().stream()
```

Comment:
Would it make more sense to just stream the keys and do a `Map.get()` operation in the filter, rather than streaming the entries and then mapping back to the key?

Reply:
For every key, we will eventually need the corresponding value for use during the `filter` stage of this code. If we only stream the keys, then we must use `Map.get()` to fetch each corresponding value. We end up doing an sweep through the map keys plus _k_ look-ups for a _k_-entry map.

In my proposed change, we are streaming the entries rather than the keys, so each item in the stream is already a (key, value) pair. What you describe as “mapping back to the key” is really just extracting the first component of such a pair. That’s going to be quite efficient, as it does not require an additional map lookup.

I am marking this as resolved, but please reopen and clarify if you still think that there is a good reason to stream keys instead of (key, value) pairs.